### PR TITLE
gprecoverseg: rebalance all possible segments

### PR DESCRIPTION
**Pull request by kalensk**
Don't bail out early if some segments can not be rebalanced. Rather rebalance all possible segments, and skip those that are down. Down segments can occur for hosts that are unreachable.

Rebalance segment pairs that are swapped, status is up, host is reachable, and synchronized.

[Pipeline](https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/7X_rebalance)

### CODE REVIEW

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 1]
```diff
@@ -2960,3 +2960,15 @@ def impl(context, including):
             raise Exception("no mirror found for segPair: %s" % segPair)
         if not are_on_different_subnets(segPair.primaryDB.hostname, segPair.mirrorDB.hostname):
             raise Exception("segmentPair on same subnet: %s" % segPair)
+
+
+@then('on content {} the primary and mirror in their {} roles')
```
Comment (jmcatamney) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:2965]: Nit: This should be "...primary and mirror *are* in...".

File: gpMgmt/bin/gppylib/gparray.py [hunk 2]
```diff
@@ -447,6 +447,19 @@ def get_hosts(self):
             hosts.append(self.mirrorDB.hostname)
         return hosts
 
+    def swapped(self):
+        return self.primaryDB.preferred_role != self.primaryDB.role and \
+               self.mirrorDB.preferred_role != self.mirrorDB.role
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gppylib/gparray.py:452]: I think it's clearer to call this function `unbalanced`.  That's standard Greenplum terminology; `swapped` is not.

File: gpMgmt/bin/gppylib/operations/rebalanceSegments.py [hunk 3]
```diff
@@ -32,12 +32,23 @@ def __init__(self, gpEnv, gpArray):
...
+        for segmentPair in self.gpArray.segmentPairs:
+            if not segmentPair.swapped():
+                continue
+
+            if segmentPair.up() and segmentPair.reachable() and segmentPair.synchronized():
+                unbalanced_primary_segs.append(segmentPair.primaryDB)
+            else:
+                self.logger.warning(
+                    "Not rebalancing primary segment dbid %d with its mirror dbid %d because one is either down, unreachable, or not synchronized" \
+                    % (segmentPair.primaryDB.dbid, segmentPair.mirrorDB.dbid))
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gppylib/operations/rebalanceSegments.py:48]: It's probably cleaner to check for not having any segments we can rebalance here and return early:

```python
if not len(unbalanced_primary_segs):
    self.logger.info("No segments can be rebalanced")
    return True
```

**Review by jmcatamney — APPROVED** [review]
LGTM other than the one nitpick.

File: gpMgmt/bin/gppylib/programs/clsRecoverSegment.py [hunk 4]
```diff
@@ -592,17 +592,6 @@ def run(self):
...
-        # Acceptable state is:
-        #    - No segments down
-        #    - No segments in change tracking or unsynchronized state
-        if self.__options.rebalanceSegments:
-            if len(gpArray.get_invalid_segdbs()) > 0:
-                raise Exception("Down segments still exist.  All segments must be up to rebalance.")
-            if len(gpArray.get_synchronized_segdbs()) != len(gpArray.getSegDbList()):
-                raise Exception(
-                    "Some segments are not yet synchronized.  All segments must be synchronized to rebalance.")
-
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gppylib/programs/clsRecoverSegment.py:605]: A check of this form should likely stay somewhere in the code path for rebalance.  We [document](https://gpdb.docs.pivotal.io/6-13/utility_guide/ref/gprecoverseg.html) that `All segments must be valid and synchronized before running gprecoverseg -r` and this is the check for that.  It's more fitting for it to be in `GpSegmentRebalanceOperation::rebalance()`.

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature [hunk 5]
```diff
@@ -174,6 +174,14 @@ Feature: gprecoverseg tests
       And the status of the primary on content 0 should be "u"
       And the status of the primary on content 1 should be "d"
 
+      # Rebalance all possible segments.
+      # Skip rebalancing mirrors that are down such as those that are unreachable.
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature:178]: nit:  Consider a clearer, test-specific phrasing like:
```python
      # Rebalance content 0 as it is still reachable.
      # Skip rebalancing content 1 as it is not reachable.
```

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature [hunk 6]
```diff
@@ -174,6 +174,14 @@ Feature: gprecoverseg tests
...
+      When the user runs "gprecoverseg -ar"
+      Then gprecoverseg should return a return code of 0
+      Then gprecoverseg should print "Not rebalancing primary segment dbid \d with its mirror dbid \d because one is either down, unreachable, or not synchronized" to stdout
+      And on content 0 the primary and mirror in their preferred roles
+      And on content 1 the primary and mirror in their non-preferred roles
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature:183]: Maybe reword this step as

```gerkin
   And content 0 is rebalanced.
   And content 1 is unbalanced.
```

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature [hunk 7]
```diff
@@ -174,6 +174,14 @@ Feature: gprecoverseg tests
       And the status of the primary on content 0 should be "u"
       And the status of the primary on content 1 should be "d"
 
+      # Rebalance all possible segments.
+      # Skip rebalancing mirrors that are down such as those that are unreachable.
+      When the user runs "gprecoverseg -ar"
+      Then gprecoverseg should return a return code of 0
+      Then gprecoverseg should print "Not rebalancing primary segment dbid \d with its mirror dbid \d because one is either down, unreachable, or not synchronized" to stdout
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gprecoverseg.feature:181]: nit: s/Then/And

**Review by dkrieger-pvtl — COMMENTED** [review]
Very clean implementation, but a few changes are suggested.

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 1 above)
Comment (kalensk) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Updated to `And content 0 is rebalanced.` per @dkrieger-pvtl  comment below.

File: gpMgmt/bin/gppylib/gparray.py [hunk 8]
```diff
@@ -447,6 +447,19 @@ def get_hosts(self):
             hosts.append(self.mirrorDB.hostname)
         return hosts
 
+    def balanced(self):
+        return self.primaryDB.preferred_role == self.primaryDB.role and \
+               self.mirrorDB.preferred_role == self.mirrorDB.role
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gppylib/gparray.py:452]: I think we want to use:

```python
    def unbalanced(self):
        return self.primaryDB.preferred_role != self.primaryDB.role and \
               self.mirrorDB.preferred_role != self.mirrorDB.role
```

and make the corresponding check below:

```python
    if not segPair.unbalanced():
        continue
```

The reason is that there are 4 cases for role/preferred role for the primary/mirror: they can both be unequal, both equal or one of each.  It is the "one of each" cases(primary down, hence it's pref_role and role are both 'p' and the mirror has been promoted) case that makes me want to use `not unbalanced()` here.

File: gpMgmt/bin/gppylib/programs/clsRecoverSegment.py (hunk 4 above)
Comment (dkrieger-pvtl) [reply gpMgmt/bin/gppylib/programs/clsRecoverSegment.py:605]: Resolved...see comment added in this PR in recover().

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 9]
```diff
@@ -2960,3 +2960,18 @@ def impl(context, including):
...
             raise Exception("segmentPair on same subnet: %s" % segPair)
+
+
+@then('content {} is {}')
+def impl(context, content, desired_state):
+    acceptable_states = ["balanced", "unbalanced"]
+    if desired_state not in acceptable_states:
+        raise Exception("expected desired state to be one of %s", acceptable_states)
+
+    role_operator = "=" if desired_state == "unbalanced" else "<>"
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:2971]: shouldn't the logic here be swapped?  That is, preferred_role `=` role if `desired_state == balanced`?

**Review by dkrieger-pvtl — COMMENTED** [review]
A few more comments...let's discuss.

File: gpMgmt/bin/gppylib/gparray.py (hunk 8 above)
Comment (dkrieger-pvtl) [reply gpMgmt/bin/gppylib/gparray.py:452]: `unbalanced` and `balanced` are equivalent IF we are guaranteed that any read of gp_segment_configuration returns either preferred_role == role or preferred_role != role for BOTH the primary and mirror.  It turns out this is true.

I'd at least add a comment at the top of this function stating this.

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 10]
```diff
@@ -2960,3 +2960,18 @@ def impl(context, including):
             raise Exception("no mirror found for segPair: %s" % segPair)
         if not are_on_different_subnets(segPair.primaryDB.hostname, segPair.mirrorDB.hostname):
             raise Exception("segmentPair on same subnet: %s" % segPair)
+
+
+@then('content {} is {}')
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:2965]: Perhaps we should rename this...it could be "anything".  Even though it's limited in use now.  Maybe 

```gherkin
@then(`content {} is {} balanced`)
```

**Review by dkrieger-pvtl — APPROVED** [review]
LGTM...provided we are sure that the Behave test we were changing last night actually is doing its job.  But consider:

1). adding a comment above `unbalanced`(see PR comment)
2). renaming the Behave step(see PR comment)
